deepmd/infer/deep_eval.py

```diff
@@ -0,0 +1,497 @@
+    Any,
+    ClassVar,
+    Dict,
+    List,
+    Optional,
+    Tuple,
+    Union,
+)
+
+import numpy as np
+
+from deepmd.dpmodel.output_def import (
+    FittingOutputDef,
+    ModelOutputDef,
+)
+from deepmd.utils.batch_size import (
+    AutoBatchSize,
+)
+
+from .backend import (
+    DPBackend,
+    detect_backend,
+)
+
+if TYPE_CHECKING:
+    import ase.neighborlist
+
+
+class DeepEvalBackend(ABC):
+    """Low-level Deep Evaluator interface.
+
+    Backends should inherbit implement this interface. High-level interface
+    will be built on top of this.
+
+    Parameters
+    ----------
+    model_file : Path
+        The name of the frozen model file.
+    *args : list
+        Positional arguments.
+    auto_batch_size : bool or int or AutoBatchSize, default: True
+        If True, automatic batch size will be used. If int, it will be used
+        as the initial batch size.
+    neighbor_list : ase.neighborlist.NewPrimitiveNeighborList, optional
+        The ASE neighbor list class to produce the neighbor list. If None, the
+        neighbor list will be built natively in the model.
+    **kwargs : dict
+        Keyword arguments.
+    """
+
+    _OUTDEF_DP2BACKEND: ClassVar[dict] = {
+        "energy": "atom_energy",
+        "energy_redu": "energy",
+        "energy_derv_r": "force",
+        "energy_derv_c": "atom_virial",
+        "energy_derv_c_redu": "virial",
+        "polar": "polar",
+        "polar_redu": "global_polar",
+        "polar_derv_r": "force",
+        "polar_derv_c": "atom_virial",
+        "polar_derv_c_redu": "virial",
+        "dipole": "dipole",
+        "dipole_redu": "global_dipole",
+        "dipole_derv_r": "force",
+        "dipole_derv_c": "atom_virial",
+        "dipole_derv_c_redu": "virial",
+        "dos": "atom_dos",
+        "dos_redu": "dos",
+    }
+
+    @abstractmethod
+    def __init__(
+        self,
+        model_file: str,
+        output_def: ModelOutputDef,
+        *args: List[Any],
+        auto_batch_size: Union[bool, int, AutoBatchSize] = True,
+        neighbor_list: Optional["ase.neighborlist.NewPrimitiveNeighborList"] = None,
+        **kwargs: Dict[str, Any],
+    ) -> None:
+        pass
+
+    def __new__(cls, model_file: str, *args, **kwargs):
+        if cls is DeepEvalBackend:
+            backend = detect_backend(model_file)
+            if backend == DPBackend.TensorFlow:
+                from deepmd.tf.infer.deep_eval import DeepEval as DeepEvalTF
+
+                return super().__new__(DeepEvalTF)
+            elif backend == DPBackend.PyTorch:
+                from deepmd.pt.infer.deep_eval import DeepEval as DeepEvalPT
+
+                return super().__new__(DeepEvalPT)
+            else:
+                raise NotImplementedError("Unsupported backend: " + str(backend))
+        return super().__new__(cls)
+
+    @abstractmethod
+    def eval(
+        self,
+        coords: np.ndarray,
+        cells: np.ndarray,
+        atom_types: List[int],
```

Comment:
Here atom_types should be np.ndarray? Because it's after DeepEvalWrapper's standardization.